src/test/perl/path.t

```diff
@@ -277,63 +310,271 @@ $mc->{fail} = undef;
 
-ok(symlink("really_really_missing", $brokenlink), "broken symlink created");
-makefile("$basetest/tgtdir/tgtfile");
-ok(symlink("tgtdir", $dirlink), "directory symlink created");
-ok(symlink("tgtdir/tgtfile", $filelink), "file symlink created");
+# Function to do the ok()/is() pair, taking into account NoAction flag
+sub check_symlink {
+    my ($mc, $target, $link_path, $expected_status) = @_;
+    my $noaction = $mc->_get_noaction();
+
+    my $ok_msg;
+    if ( $noaction ) {
+        $ok_msg = "$link_path symlink not created (NoAction set)";
+    } else {
+        $ok_msg = "$link_path is " . ($expected_status == CHANGED ? "" : "already ") . "a symlink";
+    }
+    my $target_msg = "$link_path symlink has the expected " . ($expected_status == CHANGED ? "changed " : "") . "target ($target)";
+
+    # Test if symlink was created or not according to NoAction flag
+    my $ok_condition;
+    if ( $noaction ) {
+        $ok_condition = ! $mc->any_exists($link_path);
+    } else {
+        $ok_condition = $mc->is_symlink($link_path);
+    }
+    ok($ok_condition, $ok_msg);
+
+    is(readlink($link_path), $target, $target_msg) unless $noaction;
+};
+
+init_exception("symlink tests");
+
+rmtree ($basetest) if -d $basetest;
+my $target_directory = "tgtdir";
+my $target_file1 = "tgtfile1";
+my $target_file2 = "$basetest/$target_directory/tgtfile2";
+makefile("$basetest/$target_file1");
+makefile($target_file2);
+my %opts;
+
+# Symlink creations
+for $CAF::Object::NoAction (1,0) {
+    is($mc->symlink($target_directory, $dirlink), CHANGED, "directory symlink created");
+    check_symlink($mc, $target_directory, $dirlink, CHANGED);
+    is($mc->symlink($target_file2, $filelink), CHANGED, "file symlink created");
+    check_symlink($mc, $target_file2, $filelink, CHANGED);
+}
+
+# Valid symlink updates
+# The following tests make no sens if NoAction is true (require the previous creations to occur)
+is($mc->symlink($target_file2, $filelink), SUCCESS, "file symlink already exists: nothing done");
+check_symlink($mc, $target_file2, $filelink, SUCCESS);
+is($mc->symlink($target_file1, $filelink), CHANGED, "file symlink updated");
+check_symlink($mc, $target_file1, $filelink, CHANGED);
+# Check that the file to be redefined as a symlink really exists
+ok(! readlink($target_file2), "$target_file2 exists and is a file");
+my $link_status = $mc->symlink($target_file1, $target_file2);
+ok(! defined($link_status), "symlink failed: existing file not replaced by a symlink");
+is($mc->{fail},
+   "*** cannot symlink $target_file2: it is not an existing symlink",
+   "fail attribute set after symlink failure (existing file not replaced by a symlink)");
+ok($mc->file_exists($target_file2) && ! $mc->is_symlink($target_file2), "File $target_file2 has not be replaced by a symlink");
+$opts{force} = 1;
+$CAF::Object::NoAction = 1;
+is($mc->symlink($target_file1, $target_file2, %opts), CHANGED, "existing file replaced by a symlink (force option set)");
```

Comment:
adapt message. this is force=1, but also NoAction=1; so the file shouldn't be replaced.